etw/rust/tracelogging_dynamic/src/provider.rs

```diff
@@ -183,32 +172,93 @@ impl Provider {
         return ProviderOptions::new();
     }
 
-    /// Creates a new unregistered provider.
+    /// Creates a new unregistered provider using `Guid::from_name(name)` as the provider
+    /// id.
     ///
     /// Use `register()` to register the provider. If the provider is not registered,
     /// `enabled()` will return false and `EventBuilder::write()` will be a no-op.
-    pub const fn new() -> Self {
+    ///
+    /// `name` must be less than 32KB and must not contain `'\0'`. It should be short,
+    /// human-readable, and unique enough to not conflict with names of other providers.
+    /// The provider name will typically include a company name and a component name,
+    /// e.g. "MyCompany.MyComponent".
+    ///
+    /// `options` can usually be `&Provider::options()`. If the provider needs to
+    /// join a provider group, use `Provider::options().group_id(provider_group_id)`.
+    /// If the provider needs to specify a custom provider enable callback, use
+    /// `Provider::options().callback(callback_fn, callback_context)`.
+    pub fn new(name: &str, options: &ProviderOptions) -> Self {
```

Comment:
Note: Primary change - replace `new()` with `new(args...)`, and replace `register(args...)` with `register()`. Instead of configuring during register, we configure during construction, and that allows the register operation to be an immutable operation, and that means the object has no mutable operations, making threading much easier.